Refactor Admin Member Detail view - Tools section

rhymes opened the pull request:
<!--
     For Work In Progress Pull Requests, please use the Draft PR feature,
     see https://github.blog/2019-02-14-introducing-draft-pull-requests/ for further details.

     For a timely review/response, please avoid force-pushing additional
     commits if your PR already received reviews or comments.

     Before submitting a Pull Request, please ensure you've done the following:
     - 📖 Read the Forem Contributing Guide: https://github.com/forem/forem/blob/main/CONTRIBUTING.md#create-a-pull-request.
     - 📖 Read the Forem Code of Conduct: https://github.com/forem/forem/blob/main/CODE_OF_CONDUCT.md.
     - 👷‍♀️ Create small PRs. In most cases this will be possible.
     - ✅ Provide tests for your changes.
     - 📝 Use descriptive commit messages.
     - 📗 Update any related documentation and include any relevant screenshots.

     NOTE: Pull Requests from forked repositories will need to be reviewed by
     a Forem Team member before any CI builds will run. Once your PR is approved
     with a `/ci` reply to the PR, it will be allowed to run subsequent builds without
     manual approval.
-->

## What type of PR is this? (check all applicable)

- [x] Refactor
- [ ] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

Refactor Admin Member Detail view - Tools section.

The goal of this PR is threefold:

- use GitHub's ViewComponent to isolate areas as components
- use Rails's builting Ajax support to be able to fetch server rendered HTML components via JS
- refactor some of the current Admin User tools into a new view

The fourth goal (experiment with layout components) has been postponed for a following PR.

I believe this is a viable route, even though we haven't fully explored all the capabilities of view components.

## Related Tickets & Document

Part of https://github.com/forem/rfcs/pull/183

## QA Instructions, Screenshots, Recordings

1. Open the console
2. Load the user you use to navigate the Admin
3. execute `FeatureFlag.enable(:new_admin_members, user)`
4. Go to http://localhost:3000/admin/users and click on any user in the list
5. Explore the various boxes and their functionalities

### UI accessibility concerns?

Some concerns related to page transitions correctly announced to the screenreader, given that we don't have URLs for the various states.

## Added/updated tests?

- [x] Yes
- [ ] No, and this is why: _please replace this line with details on why tests
      have not been included_
- [ ] I need help with writing tests

## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

- [ ] I've updated the [Developer Docs](https://docs.forem.com) and/or
      [Admin Guide](https://admin.forem.com/), or
      [Storybook](https://storybook.forem.com/) (for Crayons components)
- [ ] I've updated the README or added inline documentation
- [ ] I will share this change in a [Changelog](https://forem.dev/t/changelog)
      or in a [forem.dev](http://forem.dev) post
- [x] I will share this change internally with the appropriate teams
- [ ] I'm not sure how best to communicate this change and need help
- [ ] This change does not need to be communicated, and this is why not: _please
      replace this line with details on why this change doesn't need to be
      shared_

rhymes reviewed (commented):
I know there's a lot going on in this PR but I tried to document it as much as possible :-)

I'm here for all the questions and all the reviews

jgaskins reviewed (commented):
As is tradition, I just have a TIL, a nitpick, and a style question. :joy:

jgaskins commented:
Also, just to understand the intent here, the PR says that it's a refactor, but this looks more like a ground-up rewrite for this feature. I'm not passing judgement on refactor vs rewrite in that statement, to be clear, but a refactor is pretty small in scope, such as a single expression or method, and doesn't change behavior at all, so it can be done without changing high-level tests.

citizen428 commented on app/components/admin/users/tools/credits_component.rb:
```diff
@@ -0,0 +1,15 @@
+  module Users
+    module Tools
+      class CreditsComponent < ViewComponent::Base
+        delegate :orgs_with_credits, to: :helpers
```
Are helpers still available within a component's view like they are in normal Rails views?

rhymes replied:
Yes they are!

citizen428 replied:
So the main reason for defining delegates here is to make it a bit more explicit which helpers we are planning on using but nothing stops us from using any other helpers in the view? We should probably try to develop a best practice around this so it doesn't end up being too inconsistent (some people delegate everything, others nothing, etc).

citizen428 commented on app/components/admin/users/tools/notes_component.rb:
```diff
@@ -0,0 +1,12 @@
+      class NotesComponent < ViewComponent::Base
+        def initialize(user:)
+          @user = user
+          @notes = user.notes.includes(:author).order(created_at: :desc).limit(10)
```
This is totally unrelated to this PR but I've been wondering for a long time if we have to many AR chains in our controllers/services and should start leveraging named scopes more. Something like

```ruby
@notes = Notes.most_recent_for(user: user, limit: 10)
```
reads much friendlier.

rhymes replied:
I agree but I think I'll leave it like this for now, in the sense that I would like to have something like that being a focused PR that changes names and refactors the code, so that it becomes intentional for the entire team

citizen428 replied:
Yes, sorry, this was one of my tangents, not a suggestion to address this here. But this certainly seems like something we could slowly introduce during downtimes, one long AR method chain at a time 😃

citizen428 commented on app/controllers/admin/users/tools/credits_controller.rb:
```diff
@@ -0,0 +1,87 @@
+        def show
+          user = ::User.find(params[:user_id])
+
+          render CreditsComponent.new(user: user), content_type: "text/html"
```
This seems like it would be easy to forget, we may want to eventually wrap this up in a helper in `Admin::ApplicationController`, something like this (completely untested):

```ruby
def render_component(component_class, *args, **kwargs)
  render component_class.new(*args, **kwargs), content_type: "text/html"
end
```

Then we call it like this:

```ruby
render_component CreditsComponent, user: user
```

The above feels quite Rails-y to me and eliminates a whole possible error class.

citizen428 commented on app/models/user.rb:
```diff
@@ -561,6 +560,24 @@ def subscribed_to_email_follower_notifications?
     notification_setting.email_follower_notifications
   end
 
+  def last_verification_date
+    email_authorizations
+      .where.not(verified_at: nil)
+      .order(created_at: :desc)
+      .first
+      &.verified_at
+  end
+
+  def reports
+    reporter_feedback_messages.or(affected_feedback_messages).or(offender_feedback_messages)
+  end
+
+  def related_negative_reactions
+    Reaction.article_vomits.where(reactable_id: article_ids)
+      .or(Reaction.comment_vomits.where(reactable_id: comment_ids))
+      .or(Reaction.user_vomits.where(user_id: id))
+  end
```
Since `User` is already very big I'm always a bit hesitant to add more to it. Is there any chance these could be scopes/class methods on the associations (`EmailAuthorization` and `Reaction`) instead where pass in a user?

rhymes replied:
Good idea, will see if they feel "Railsy" :D

citizen428 reviewed (commented):
Great effort, @rhymes! 👏 I have to say the sometimes very long `respond_to` blocks certainly take some getting used to.

aitchiss commented on app/javascript/admin/controllers/snackbar_controller.js:
```diff
@@ -0,0 +1,31 @@
+    render(null, this.snackZoneTarget);
+  }
+
+  // Any Stimulus controller can add an item to the Snackbar by dispatching a custom event
+  // data-action="snackbar:add@document->snackbar#addItem"
```
that's cool 😎

aitchiss commented on app/javascript/admin/controllers/snackbar_controller.js:
```diff
@@ -0,0 +1,31 @@
+    const { message, addCloseButton = false } = event.detail;
+
+    // eslint-disable-next-line import/no-unresolved
+    const { addSnackbarItem } = await import('Snackbar');
```
same here re caching the imports from `connect()`

aitchiss commented on app/views/admin/users/new/show.html.erb:
```diff
@@ -0,0 +1,53 @@
+  data-action="user:tools@document->user#fetchAndOpenTools">
+  <div class="flex flex-column" style="flex-basis: 70%;">
+    <section class="crayons-card" id="profile">
+      <h3 class="crayons-subtitle-2"><%= @user.name %></h3>
```
It's more of a design thing, but I think the heading hierarchy is a bit weird here. The biggest heading is "Members" when really the context of the current page is the specific user.

Also the `h1` for the page should really be the one that tells us where we are currently but at the moment it's "Users" up in the header 🙃

rhymes replied:
We could move the info about the current user in the h1, not sure what to keep here though...

aitchiss commented on cypress/integration/seededFlows/adminFlows/users/memberToolsSection.spec.js:
```diff
@@ -0,0 +1,256 @@
+        cy.findByRole('link', { name: username }).click();
+        cy.findByRole('link', { name: /Emails/ }).click();
+
+        cy.findAllByText(/Emails history/)
```
can we target by a role?

rhymes replied:
I couldn't find a suitable role for the `<details>` :(

aitchiss commented on app/components/admin/users/tools/credits_component.html.erb:
```diff
@@ -0,0 +1,144 @@
+                class: "my-3",
+                id: :credits_add do |f| %>
+    <fieldset>
+      <legend class="fs-base fw-medium mb-3"><%= pluralize(@user_unspent_credits_count, "available user credit") %></legend>
```
I don't think this should be the `legend` as it's more supplemental info than describing the input group

rhymes replied:
A generic div then?

aitchiss commented on app/components/admin/users/tools/credits_component.html.erb:
```diff
@@ -0,0 +1,144 @@
+      <legend class="fs-base fw-medium mb-3"><%= pluralize(@user_unspent_credits_count, "available user credit") %></legend>
+
+      <div class="crayons-field wb-2">
+        <%= f.label "Add credits", class: "crayons-field__label", id: "add-credits" %>
```
This should be the `legend` for the fieldset

aitchiss commented on app/components/admin/users/tools/credits_component.html.erb:
```diff
@@ -0,0 +1,144 @@
+      </div>
+
+      <div class="crayons-field form-group w-50">
+        <%= f.number_field "credits[count]",
+                           id: nil,
+                           in: 1...10_000,
+                           required: true,
+                           class: "crayons-textfield",
+                           "aria-labelledby": "add-credits" %>
+        <%= f.text_field "credits[note]",
+                         id: nil,
+                         placeholder: "Why are you adding these credits?",
+                         required: true,
+                         class: "crayons-textfield",
+                         "aria-labelledby": "add-credits" %>
```
At the moment, both of these fields have the same label, when they're for different things. I think we can just add an aria-label to each 🙂

rhymes replied:
Oh cool, I was trying to have them be described by the same thing but you're right, they have a different purpose

aitchiss replied:
Yeah the nice thing about having the "Add credits" bit as the fieldset's `legend` is that it'll still be linked to these inputs, just as a descriptive "here's what this group does", rather than the more granular "here's what this input does"... if that makes sense 😄

aitchiss commented on app/components/admin/users/tools/credits_component.html.erb:
```diff
@@ -0,0 +1,144 @@
+                  id: :credits_remove do |f| %>
+      <fieldset>
+        <div class="crayons-field mb-2">
+          <%= f.label "Remove credits", class: "crayons-field__label", id: "remove-credits" %>
```
Same here, i think this should be the `legend` for this fieldset

rhymes replied:
Will move!

aitchiss commented on app/components/admin/users/tools/credits_component.html.erb:
```diff
@@ -0,0 +1,144 @@
+        </div>
+
+        <div class="crayons-field form-group w-50">
+          <%= f.number_field "credits[count]",
+                             id: nil,
+                             in: 1..@user_unspent_credits_count,
+                             required: true,
+                             class: "crayons-textfield mr-1",
+                             "aria-labelledby": "remove-credits" %>
+          <%= f.text_field "credits[note]",
+                           id: nil,
+                           placeholder: "Why are you removing these credits?",
+                           required: true,
+                           class: "crayons-textfield mr-1",
+                           "aria-labelledby": "remove-credits" %>
```
and same here with needing individual aria-labels

aitchiss commented on app/components/admin/users/tools/credits_component.html.erb:
```diff
@@ -0,0 +1,144 @@
+
+      <fieldset>
+        <div class="crayons-field mb-1">
+          <%= f.label "Add credits to organizations", class: "crayons-field__label", id: "add-org-credits" %>
```
Same as above, this should be the `legend` and the inputs below should have aria-labels. Same for the "remove credits" section below

rhymes replied:
thanks!

aitchiss reviewed (commented):
This is really cool @rhymes 😎 

I've left some comments/thoughts/suggestions - I know there will be ongoing work on this though and some things might not be in scope right now, in which case feel free to ignore 🙂

rhymes commented on app/controllers/admin/application_controller.rb:
```diff
@@ -19,7 +19,7 @@ class ApplicationController < ApplicationController
     }.freeze
 
-    private
+    protected
```
Thank you for the explanation! I'll revert this :)

rhymes commented on spec/requests/admin/users/tools/credits_spec.rb:
```diff
@@ -0,0 +1,155 @@
+            note: "Test"
+          }
+        }, xhr: true
+      end.to change(user.credits, :count).by(1)
+        .and change(user.notes, :count).by(1)
```
I don't know :D, I prefer the first version visually and I think it's the one we use most often around the code

rhymes commented on app/components/admin/users/tools/credits_component.html.erb:
```diff
@@ -0,0 +1,144 @@
+  class="crayons-card p-3 js-component"
+  data-controller="users--tools--ajax"
+  data-action="ajax:success@document->users--tools--ajax#success ajax:error@document->users--tools--ajax#error">
+  <header>
```
Thanks!

rhymes commented on app/javascript/admin/controllers/snackbar_controller.js:
```diff
@@ -0,0 +1,31 @@
+  }
+
+  async disconnect() {
+    const { render } = await import('preact');
```
Great suggestion, thank you!

rhymes replied:
On another note, I'm not sure this can be done that way (controllers are mounted and unmounted, which means there's no survinving cache), unless we replicate the same mechanism inside the admin pack. I'll make a note for a future PR if that's okay

rhymes commented on app/assets/stylesheets/components/admin/users/tools_component.scss:
```diff
@@ -0,0 +1,6 @@
+  cursor: pointer;
+  text-decoration: none;
+  background-color: var(--base-10);
+  padding: var(--su-4) var(--su-4);
```
Yep I know, one side becomes four sides in CSS but I don't remember why I ended up specifying it twice :D Probably by accident, will rectify :D

rhymes commented:
> Also, just to understand the intent here, the PR says that it's a refactor, but this looks more like a ground-up rewrite for this feature. I'm not passing judgement on refactor vs rewrite in that statement, to be clear, but a refactor is pretty small in scope, such as a single expression or method, and doesn't change behavior at all, so it can be done without changing high-level tests.

@jgaskins true that, but we also don't have "Rewrite" as an option in the template, I chose the closest one hahaha

rhymes commented on app/components/admin/users/tools/organizations_component.html.erb:
```diff
@@ -0,0 +1,87 @@
+  data-controller="users--tools--ajax"
+  data-action="ajax:success@document->users--tools--ajax#success ajax:error@document->users--tools--ajax#error">
+  <header>
+    <h3><%= link_to "← Organizations Memberships", admin_user_tools_path(@user.id),
```
So, there's a way to generalize the header but it'd require restructuring most of the partials and use "layout management", so I'm going to leave that part off and only make sure that all the back links display `Users` for now :)

rhymes commented:
Hi @citizen428, @jgaskins and @aitchiss. I think I've addressed most of your feedback. I'm going to leave repetitions to a following PR on layout and caching of imports as well.

citizen428 reviewed (approved):
Thanks for all the work on this @rhymes, 💯

Ridhwana reviewed (approved):
Awesome work Rhymes!

aitchiss commented on app/components/admin/users/tools/emails_component.html.erb:
```diff
@@ -0,0 +1,74 @@
+  <div id="section-title" class="hidden">Emails section</div>
+
+  <h3>
+    <%= link_to "← Tools", admin_user_tools_path(@user.id),
+                remote: true,
+                data: { action: "ajax:success@document->user#replacePartial" },
+                class: "crayons-title fs-l" %>
```
It's a minor thing and can be swept up later, but it would be good if this link had a focus style (or really, all links in the admin section - I don't think it's specific to here 😅 )

rhymes replied:
Let me add that to all the back links!

aitchiss reviewed (approved):
Looks good to me! The focus management is working nicely ✨ 

It might be cool in a future iteration to refocus the Tools link they just visited when they click the "<- Tools" link, but I can see it being more trouble than it's worth and I think it's enough that we reset focus into the first box in the Tools list
